126: Add tests for API auth endpoints

EugeneTurkin opened the pull request:
We want to cover our REST API part of the codebase with unit-tests.

In the scope of this task we need to cover authentication/authorization part of the REST API code.

birthdaysgift commented on tests/conftest.py:
```diff
@@ -28,3 +29,13 @@ def db_empty():
+def db_with_one_user(db_empty):
+    session = db_empty
+    user = User(id=1, username="test", password_hash="testtest")
+    user.set_password("testtest")
```
We should try to avoid using our application code in tests as much as possible. During the test only SUT code should be executed. It will allow us to know exactly that our SUT is fine/broken depending on results of our test results - success/failure.

So here it's better to have something like this:
```python
session = db_empty
user = User(id=1, username="test")
user.password_hash = hashlib.sha256("testtest".encode()).hexdigest()
session.add(user)
session.commit()
return session
```

birthdaysgift commented on tests/auth/test_auth_endpoint.py:
```diff
@@ -0,0 +1,23 @@
+
+def test_post_auth_endpoint_with_existing_user_returns_correct_body(client, db_with_one_user):
+    session = db_with_one_user
+    user = session.query(User).first()
```
You don't need to query user here, because you already know its id, since you provided `id=1` in the fixture `db_with_one_user`. So you can use it directly:
```python
str.encode(jwt.encode({"user_id": 1}, Config.SECRET_KEY, algorithm="HS256"))
```

birthdaysgift commented on tests/auth/test_auth_endpoint.py:
```diff
@@ -0,0 +1,23 @@
+def test_post_auth_endpoint_with_nonexistent_user_returns_400_with_correct_message(client, db_empty):
+    result = client.post("/api/auth", data={"username": "test", "password": "testtest"})
+
+    assert result.status_code == 400 and b"couldn't find user with credentials provided" in result.data
```
Let's separate these assertion statements. It will allow us to know exactly what's wrong when one of them will fail.

Also it's better to assert the full response body whenever it's possible to be sure that all response is fine.

So I propose these assertions:
```python
assert result.status_code == 400
assert result.json == {"message": "couldn't find user with credentials provided"}
```

birthdaysgift commented on tests/auth/test_auth_endpoint.py:
```diff
@@ -0,0 +1,23 @@
+    result = client.post("/api/auth", data={"username": "test", "password": "testtest"})
+
+    assert result.status_code == 200
+    assert expected_hash in result.data
```
Since it's better to check the whole response body whenever it's possible, let's change this assertion to something like:
```python
assert result.json == jwt.encode({"user_id": 1}, Config.SECRET_KEY, algorithm="HS256")
```

EugeneTurkin committed:
126: Add tests for api/auth endpoints

We want to cover our REST API part of the codebase with unit-tests.

In the scope of this task we need to cover authentication/authorization part of the REST API code.
